### DISCUSSION TITLE
How to use tonic GRPC client in Dioxus？

### DISCUSSION THREAD

**Discussion by nanlong** [DioxusLabs/dioxus#437] [Q&A]
Unfortunately, I couldn't find a solution or sample code.

Is it feasible?

And if so, how?

Thank you very much.

**Comment by rMazeiks** [comment]
Hi!

I'm not very familiar with Tonic, but looks like it's an async library.

# To make a request when a component renders

You can use [`use_future`](https://docs.rs/dioxus-hooks/latest/dioxus_hooks/fn.use_future.html) and put whatever async code you need in there. That will return a [UseFuture](https://docs.rs/dioxus-hooks/latest/dioxus_hooks/struct.UseFuture.html) struct, and you can call [`.value()`](https://docs.rs/dioxus-hooks/latest/dioxus_hooks/struct.UseFuture.html#method.value) to get the result of the request (once it resolves).

As for sample code, you can take a look at Dioxus examples that have `use_future` – for example, the [dog_app](https://github.com/DioxusLabs/dioxus/blob/ff81f1c8192f9774bb144bb78d257111f82f6ff4/examples/dog_app.rs). It uses `reqwest`, not `tonic`, but the idea is similar.

# To make a request in response to a user action

You can use [cx.spawn](https://docs.rs/dioxus/latest/dioxus/prelude/struct.ScopeState.html#method.spawn) to spawn a future imperatively. An example where this is used is [login_form](https://github.com/DioxusLabs/dioxus/blob/07bea88fcc51d0c1285683e21453800c4428fd59/examples/login_form.rs). Again, `reqwest`, not `tonic`, but you can adapt it to your needs.

---

Apart from that, I don't think there will be anything Dioxus-specific, so you can just follow the usual Tonic docs. Admittedly, at first glance, they're not very simple, but it looks like they have [a nice tutorial](https://github.com/hyperium/tonic/blob/master/examples/helloworld-tutorial.md) to get started. Also take a look at [tonic examples](https://github.com/hyperium/tonic/tree/master/examples/src). Good luck!

**Comment by kalzoo** [comment]
I can add a little more here, in case others stumble on this in the future.

The key problem with `tonic` is the use of `http2` to enable gRPC. `http2` is fundamentally browser-incompatible. You can still use it, but the target gRPC service also needs to be adapted to accept http/1 requests with the grpc-web protocol.

There are a few developing libraries out there to help with this. one, for example, is https://docs.rs/tonic-web-wasm-client/latest/tonic_web_wasm_client/ ; I have not used it, YMMV.
